File: web/src/lib/swing.ts
import type { BadgeTone } from '../ui';
import { toNumber } from '../ui';

/** Plan M11.6: the swing book as the API returns it (docs/api.md, "Swing book"). Prices are JSON numbers or strings. */
export interface SwingBookRow {
  id: string; instrumentId: string; symbol: string; strategyId?: string | null; openedAt: string; entryDate: string; daysHeld: number;
  quantity: number; entryPrice: number | string; stop?: number | string | null; goal?: number | string | null; lastPrice: number | string;
  r?: number | string | null; unrealizedPnl: { paise: number }; gtt: 'ACTIVE' | 'MISSING' | 'NONE' | string; gttId?: string | null;
}

export interface SwingSetup {
  deploymentId: string; baseId: string; instrumentId: string; symbol: string; type: string; pivot: number | string; buyHigh: number | string;
  stop: number | string; goal: number | string; avgVolume50?: number | null; state: string; lastClose?: number | string | null;
  pace?: number | string | null; signalId?: string | null;
}

export interface SwingPositionRisk {
  instrumentId: string; symbol: string; quantity: number; price: number | string; stop?: number | string | null; industry?: string | null;
  risk: { paise: number };
}

export interface SwingRisk {
  mode: string; overnightRisk: { paise: number }; budget: { paise: number }; gapAllowancePct: number | string; openPositions: number;
  maxOpenPositions: number; deployed: { paise: number }; capital: { paise: number }; positions: SwingPositionRisk[];
}

/** A missing broker-side stop is an incident; no GTT yet (the seconds after a fill) a warning. */
export function gttTone(state: string): BadgeTone {
  return state === 'ACTIVE' ? 'profit' : state === 'MISSING' ? 'loss' : 'warning';
}

/** The GTT state with a mark, so colour is never the only signal. */
export function gttLabel(state: string): string {
  return state === 'ACTIVE' ? '✓ GTT active' : state === 'MISSING' ? '✗ GTT missing' : '⚠ no GTT';
}

/** The watcher's state of a setup (docs/swing.md): triggered is the news, the refusals are warnings. */
export function setupTone(state: string): BadgeTone {
  switch (state) {
    case 'TRIGGERED': return 'profit';
    case 'ABOVE_BUY_ZONE': case 'NO_VOLUME': case 'STOP_TOO_NEAR': return 'warning';
    default: return 'neutral';
  }
}

export function setupLabel(state: string): string {
  return state.toLowerCase().replace(/_/g, ' ');
}

/** The share of the overnight budget in use, and its tone: over budget is a loss, three quarters a warning. */
export function riskUse(used: number, budget: number): { pct: number | null; tone: BadgeTone } {
  if (!budget || budget <= 0) return { pct: null, tone: 'neutral' };
  const pct = Math.round((1000 * used) / budget) / 10;
  return { pct, tone: pct > 100 ? 'loss' : pct >= 75 ? 'warning' : 'profit' };
}

/** A price for display: two decimals, `—` for none. */
export function price(v: unknown): string {
  const n = toNumber(v);
  return n === null ? '—' : n.toFixed(2);
}

File: tui/internal/api/swing.go
package api

import "net/http"

// The swing book (plan Phase 11): delivery positions held overnight, protected by a GTT at the broker. PAPER only.

// SwingPosition is one open position of the swing book (GET /swing/positions).
type SwingPosition struct {
	ID            string `json:"id"`
	InstrumentID  string `json:"instrumentId"`
	Symbol        string `json:"symbol"`
	EntryDate     string `json:"entryDate"`
	DaysHeld      int    `json:"daysHeld"`
	Quantity      int    `json:"quantity"`
	EntryPrice    Flex   `json:"entryPrice"`
	Stop          *Flex  `json:"stop"`
	Goal          *Flex  `json:"goal"`
	LastPrice     Flex   `json:"lastPrice"`
	R             *Flex  `json:"r"`
	UnrealizedPnl Paise  `json:"unrealizedPnl"`
	Gtt           string `json:"gtt"`
	GttID         string `json:"gttId"`
}

// SwingSetup is one READY trade plan the swing watcher watches today (GET /swing/setups).
type SwingSetup struct {
	Symbol    string `json:"symbol"`
	Type      string `json:"type"`
	Pivot     Flex   `json:"pivot"`
	BuyHigh   Flex   `json:"buyHigh"`
	Stop      Flex   `json:"stop"`
	Goal      Flex   `json:"goal"`
	State     string `json:"state"`
	LastClose *Flex  `json:"lastClose"`
	Pace      *Flex  `json:"pace"`
	SignalID  string `json:"signalId"`
}

// SwingPositionRisk is one position's gap-adjusted overnight risk.
type SwingPositionRisk struct {
	Symbol   string `json:"symbol"`
	Quantity int    `json:"quantity"`
	Price    Flex   `json:"price"`
	Stop     *Flex  `json:"stop"`
	Industry string `json:"industry"`
	Risk     Paise  `json:"risk"`
}

// SwingRisk is the swing book's overnight risk against its budget (GET /swing/risk).
type SwingRisk struct {
	Mode             string              `json:"mode"`
	OvernightRisk    Paise               `json:"overnightRisk"`
	Budget           Paise               `json:"budget"`
	GapAllowancePct  Flex                `json:"gapAllowancePct"`
	OpenPositions    int                 `json:"openPositions"`
	MaxOpenPositions int                 `json:"maxOpenPositions"`
	Deployed         Paise               `json:"deployed"`
	Capital          Paise               `json:"capital"`
	Positions        []SwingPositionRisk `json:"positions"`
}

func (c *Client) SwingPositions() ([]SwingPosition, error) {
	var p []SwingPosition
	return p, c.do(http.MethodGet, "/swing/positions", nil, false, &p)
}

func (c *Client) SwingSetups() ([]SwingSetup, error) {
	var s []SwingSetup
	return s, c.do(http.MethodGet, "/swing/setups", nil, false, &s)
}

func (c *Client) SwingRisk() (SwingRisk, error) {
	var r SwingRisk
	return r, c.do(http.MethodGet, "/swing/risk", nil, false, &r)
}

// SwingClose exits one swing position by symbol; its GTT is cancelled in the same operation.
func (c *Client) SwingClose(symbol string) (map[string]any, error) {
	var out map[string]any
	return out, c.do(http.MethodPost, "/swing/positions/close", map[string]string{"symbol": symbol}, true, &out)
}
